File: backend/api/messages.py
"""Лента уровня 2: список сообщений источника, детали, прочитано, ручной уровень."""
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from backend.db import get_db
from backend.events import bus
from backend.models import Message, OverrideLog, User
from backend.queries import apply_filters, user_messages
from backend.schemas import LevelUpdate, MessageList, MessageOut
from backend.security import current_user
from backend.serializers import message_out

router = APIRouter(prefix="/api/messages", tags=["messages"])


def _get_owned(db: Session, user: User, message_id: int) -> Message:
    message = db.get(Message, message_id)
    if message is None or message.connection.user_id != user.id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Сообщение не найдено")
    return message


@router.get("", response_model=MessageList)
def list_messages(
    source: str | None = Query(default=None, pattern="^(gmail|telegram)$"),
    level: str = Query(default="all", pattern="^(all|CRITICAL|HIGH|NORMAL|LOW)$"),
    status_filter: str = Query(default="all", alias="status", pattern="^(all|unread|read|done)$"),
    reply: str = Query(default="all", pattern="^(all|yes|no)$"),
    action: str = Query(default="all", pattern="^(all|yes|no)$"),
    period: str = Query(default="all", pattern="^(all|today|week|month)$"),
    q: str | None = None,
    tz_offset: int = Query(default=0, description="смещение часового пояса клиента, минуты"),
    user: User = Depends(current_user),
    db: Session = Depends(get_db),
) -> MessageList:
    stmt = apply_filters(
        user_messages(user),
        source=source,
        level=level,
        status=status_filter,
        reply=reply,
        action=action,
        period=period,
        q=q,
        tz_offset=tz_offset,
    )
    messages = db.execute(
        stmt.order_by(Message.received_at.desc(), Message.id.desc())
    ).scalars().all()
    unread = sum(1 for message in messages if not message.is_read)
    return MessageList(
        items=[message_out(message) for message in messages],
        total=len(messages),
        unread=unread,
    )


@router.get("/{message_id}", response_model=MessageOut)
def get_message(
    message_id: int,
    user: User = Depends(current_user),
    db: Session = Depends(get_db),
) -> MessageOut:
    return message_out(_get_owned(db, user, message_id))


@router.post("/{message_id}/read", response_model=MessageOut)
def mark_read(
    message_id: int,
    user: User = Depends(current_user),
    db: Session = Depends(get_db),
) -> MessageOut:
    """Помечается при открытии деталей. Обратной операции в MVP нет (решение №12)."""
    message = _get_owned(db, user, message_id)
    if not message.is_read:
        message.is_read = True
        db.commit()
        db.refresh(message)
    return message_out(message)


@router.post("/{message_id}/level", response_model=MessageOut)
def set_level(
    message_id: int,
    payload: LevelUpdate,
    user: User = Depends(current_user),
    db: Session = Depends(get_db),
) -> MessageOut:
    """Ручное исправление уровня — единственная точка правки оценки."""
    message = _get_owned(db, user, message_id)
    previous = message.effective_level
    if payload.level != previous:
        db.add(
            OverrideLog(message_id=message.id, from_level=previous, to_level=payload.level)
        )
    message.level_override = payload.level
    db.commit()
    db.refresh(message)
    bus.publish(user.id, "message.analyzed", message_out(message).model_dump(mode="json"))
    return message_out(message)

File: backend/internal/routers/messages.go
package routers

import (
	"errors"
	"net/http"
	"personalinbox/internal/exceptions"
	"strconv"

	"personalinbox/internal/postgres"
	"personalinbox/internal/schemas"
)

// param читает значение из строки запроса и проверяет его по списку
// допустимых. Неизвестное значение — 422, как у прежнего бэкенда.
func param(w http.ResponseWriter, r *http.Request, name, fallback string, allowed []string) (string, bool) {
	value := r.URL.Query().Get(name)
	if value == "" {
		return fallback, true
	}
	if !postgres.Contains(allowed, value) {
		fail(w, http.StatusUnprocessableEntity, "Недопустимое значение фильтра «"+name+"»")
		return "", false
	}
	return value, true
}

// ownedMessage — сообщение пользователя. Чужое сообщение — 404, а не 403:
// о его существовании знать незачем.
func (s *Server) ownedMessage(w http.ResponseWriter, r *http.Request, user *postgres.User) (*postgres.Message, bool) {
	id, err := strconv.ParseInt(r.PathValue("id"), 10, 64)
	if err != nil {
		fail(w, http.StatusNotFound, "Сообщение не найдено")
		return nil, false
	}
	message, err := s.db.MessageByID(id)
	if errors.Is(err, exceptions.ErrNotFound) {
		fail(w, http.StatusNotFound, "Сообщение не найдено")
		return nil, false
	}
	if err != nil {
		fail(w, http.StatusInternalServerError, "Не удалось получить сообщение")
		return nil, false
	}
	connection, err := s.db.ConnectionByID(message.ConnectionID)
	if err != nil || connection.UserID != user.ID {
		fail(w, http.StatusNotFound, "Сообщение не найдено")
		return nil, false
	}
	return message, true
}

func (s *Server) handleListMessages(w http.ResponseWriter, r *http.Request) {
	user, ok := s.currentUser(w, r)
	if !ok {
		return
	}

	source, ok := param(w, r, "source", "", postgres.SourceKinds)
	if !ok {
		return
	}
	level, ok := param(w, r, "level", "all",
		[]string{"all", "CRITICAL", "HIGH", "NORMAL", "LOW"})
	if !ok {
		return
	}
	status, ok := param(w, r, "status", "all", []string{"all", "unread", "read", "done"})
	if !ok {
		return
	}
	reply, ok := param(w, r, "reply", "all", []string{"all", "yes", "no"})
	if !ok {
		return
	}
	action, ok := param(w, r, "action", "all", []string{"all", "yes", "no"})
	if !ok {
		return
	}
	period, ok := param(w, r, "period", "all", []string{"all", "today", "week", "month"})
	if !ok {
		return
	}

	tzOffset := 0
	if raw := r.URL.Query().Get("tz_offset"); raw != "" {
		parsed, err := strconv.Atoi(raw)
		if err != nil {
			fail(w, http.StatusUnprocessableEntity, "tz_offset должен быть числом")
			return
		}
		tzOffset = parsed
	}

	messages, err := s.db.FilteredMessages(user.ID, postgres.Filter{
		Source:   source,
		Level:    level,
		Status:   status,
		Reply:    reply,
		Action:   action,
		Period:   period,
		Q:        r.URL.Query().Get("q"),
		TZOffset: tzOffset,
	})
	if err != nil {
		fail(w, http.StatusInternalServerError, "Не удалось получить сообщения")
		return
	}

	items := make([]schemas.Message, 0, len(messages))
	unread := 0
	for _, message := range messages {
		items = append(items, schemas.MessageOut(message))
		if !message.IsRead {
			unread++
		}
	}
	respond(w, http.StatusOK, schemas.MessageList{
		Items:  items,
		Total:  len(items),
		Unread: unread,
	})
}

func (s *Server) handleGetMessage(w http.ResponseWriter, r *http.Request) {
	user, ok := s.currentUser(w, r)
	if !ok {
		return
	}
	message, ok := s.ownedMessage(w, r, user)
	if !ok {
		return
	}
	respond(w, http.StatusOK, schemas.MessageOut(message))
}

// handleMarkRead: помечается при открытии деталей. Обратной операции
// в MVP нет (решение №12).
func (s *Server) handleMarkRead(w http.ResponseWriter, r *http.Request) {
	user, ok := s.currentUser(w, r)
	if !ok {
		return
	}
	message, ok := s.ownedMessage(w, r, user)
	if !ok {
		return
	}
	if !message.IsRead {
		message.IsRead = true
		if err := s.db.SaveMessage(message); err != nil {
			fail(w, http.StatusInternalServerError, "Не удалось сохранить сообщение")
			return
		}
		// Событие в шину не уходит (карточка не перерисовывается), а счётчик
		// непрочитанных в кэше источников уже неверный — сбрасываем руками.
		s.cache.DropCache(r.Context(), user.ID)
	}
	respond(w, http.StatusOK, schemas.MessageOut(message))
}

// handleSetLevel — ручное исправление уровня, единственная точка правки оценки.
func (s *Server) handleSetLevel(w http.ResponseWriter, r *http.Request) {
	user, ok := s.currentUser(w, r)
	if !ok {
		return
	}
	message, ok := s.ownedMessage(w, r, user)
	if !ok {
		return
	}
	var payload struct {
		Level string `json:"level"`
	}
	if !readJSON(w, r, &payload) {
		return
	}
	if !postgres.Contains(postgres.Levels, payload.Level) {
		fail(w, http.StatusUnprocessableEntity, "Неизвестный уровень важности")
		return
	}

	previous := message.EffectiveLevel()
	if payload.Level != previous {
		if err := s.db.AddOverrideLog(message.ID, previous, payload.Level); err != nil {
			fail(w, http.StatusInternalServerError, "Не удалось сохранить исправление")
			return
		}
	}
	message.LevelOverride = payload.Level
	if err := s.db.SaveMessage(message); err != nil {
		fail(w, http.StatusInternalServerError, "Не удалось сохранить сообщение")
		return
	}
	s.bus.Publish(user.ID, "message.analyzed", schemas.MessageOut(message))
	respond(w, http.StatusOK, schemas.MessageOut(message))
}
